File: src/test/java/uk/gov/justice/hmpps/offendersearch/controllers/OffenderSearchAPITest.java
package uk.gov.justice.hmpps.offendersearch.controllers;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.tomakehurst.wiremock.junit.WireMockRule;
import io.restassured.RestAssured;
import io.restassured.config.ObjectMapperConfig;
import io.restassured.config.RestAssuredConfig;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import uk.gov.justice.hmpps.offendersearch.dto.OffenderDetail;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;
import static io.restassured.RestAssured.given;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles("dev,localstack")
@RunWith(SpringJUnit4ClassRunner.class)
public class OffenderSearchAPITest {

    @Rule
    public WireMockRule wireMock = new WireMockRule(wireMockConfig().port(4571).jettyStopTimeout(10000L));

    @LocalServerPort
    int port;

    @Autowired
    private ObjectMapper objectMapper;

    @Value("${test.token.good}")
    private String validOauthToken;

    @Before
    public void setup() {
        RestAssured.port = port;
        RestAssured.config = RestAssuredConfig.config().objectMapperConfig(
                new ObjectMapperConfig().jackson2ObjectMapperFactory((aClass, s) -> objectMapper));
    }

    @Test
    public void offenderSearch() throws IOException {

            stubFor(get(anyUrl()).willReturn(
                    okForContentType("application/json", response("src/test/resources/elasticsearchdata/singleMatch.json"))));


        final var results = given()
                .auth()
                .oauth2(validOauthToken)
                .contentType(APPLICATION_JSON_VALUE)
                .body("{\"name\":\"smith\"}")
                .when()
                .get("/search")
                .then()
                .statusCode(200)
                .extract()
                .body()
                .as(OffenderDetail[].class);

         assertThat(results).hasSize(1);
         assertThat(results).extracting("firstName").containsOnly("John");
    }

    private String response(String file) throws IOException {
        return Files.readString(Paths.get(file));
    }
}

File: src/test/kotlin/uk/gov/justice/hmpps/offendersearch/controllers/OffenderSearchAPITest.kt
package uk.gov.justice.hmpps.offendersearch.controllers

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.core.WireMockConfiguration
import com.github.tomakehurst.wiremock.junit.WireMockRule
import io.restassured.RestAssured
import io.restassured.config.ObjectMapperConfig
import io.restassured.config.RestAssuredConfig
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.hamcrest.CoreMatchers
import org.junit.*
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.MediaType
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner
import uk.gov.justice.hmpps.offendersearch.dto.OffenderDetail
import java.lang.reflect.Type
import java.nio.file.Files
import java.nio.file.Paths

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles("dev,wiremock")
@RunWith(SpringJUnit4ClassRunner::class)
class OffenderSearchAPITest {
  @LocalServerPort
  var port = 0
  @Autowired
  private lateinit var objectMapper: ObjectMapper
  @Value("\${test.token.good}")
  private val validOauthToken: String? = null

  companion object {
    private val wireMock = WireMockRule(WireMockConfiguration.wireMockConfig().port(4444).jettyStopTimeout(10000L))

    @BeforeClass
    @JvmStatic
    fun startMocks() {
      wireMock.start()
    }

    @AfterClass
    @JvmStatic
    fun stopMocks() {
      wireMock.stop()
    }
  }
  @Before
  fun setup() {
    RestAssured.port = port
    RestAssured.config = RestAssuredConfig.config().objectMapperConfig(
        ObjectMapperConfig().jackson2ObjectMapperFactory { _: Type?, _: String? -> objectMapper })
  }

  @Test
  fun offenderSearch() {
    wireMock.stubFor(WireMock.get(WireMock.anyUrl()).willReturn(
        WireMock.okForContentType("application/json", response("src/test/resources/elasticsearchdata/singleMatch.json"))))
    val results = RestAssured.given()
        .auth()
        .oauth2(validOauthToken)
        .contentType(MediaType.APPLICATION_JSON_VALUE)
        .body("{\"surname\":\"smith\"}")
        .`when`()["/search"]
        .then()
        .statusCode(200)
        .extract()
        .body()
        .`as`(Array<OffenderDetail>::class.java)
    assertThat(results).hasSize(1)
    assertThat(results).extracting("firstName").containsOnly("John")
  }

  @Test
  fun noSearchParameters_badRequest() {
    RestAssured.given()
        .auth()
        .oauth2(validOauthToken)
        .contentType(MediaType.APPLICATION_JSON_VALUE)
        .body("{}")
        .`when`()["/search"]
        .then()
        .statusCode(400)
        .body("developerMessage", CoreMatchers.containsString("Invalid search  - please provide at least 1 search parameter"))
  }

  @Test
  fun invalidDateOfBirthFormat_badRequest() {
    RestAssured.given()
        .auth()
        .oauth2(validOauthToken)
        .contentType(MediaType.APPLICATION_JSON_VALUE)
        .body("{\"dateOfBirth\":\"23/11/1976\"}")
        .`when`()["/search"]
        .then()
        .statusCode(400)
  }

  private fun response(file: String): String {
    return Files.readString(Paths.get(file))
  }
}